homeassistant/components/firmata/board.py

```diff
@@ -0,0 +1,183 @@
+
+        _LOGGER.info("Firmata connection successful for %s", self.name)
+        return True
+
+    async def async_reset(self):
+        """Reset the board to default state."""
+        # If the board was never setup, continue.
+        if self.api is None:
+            return True
+
+        await self.api.shutdown()
+
+        return True
+
+    async def async_update_device_registry(self):
+        """Update board registry."""
+        device_registry = await self.hass.helpers.device_registry.async_get_registry()
+        device_registry.async_get_or_create(
+            config_entry_id=self.config_entry.entry_id,
+            connections={},
+            identifiers={(DOMAIN, self.name)},
+            manufacturer="Firmata",
+            name=self.name,
+            sw_version=self.firmware_version,
+        )
+
+
+async def get_board(data: dict):
+    """Create a Pymata board object."""
+    board_data = {}
+
+    if CONF_SERIAL_PORT in data:
+        board_data["com_port"] = data[CONF_SERIAL_PORT]
+    if CONF_SERIAL_BAUD_RATE in data:
+        board_data["baud_rate"] = data[CONF_SERIAL_BAUD_RATE]
+    if CONF_ARDUINO_INSTANCE_ID in data:
+        board_data["arduino_instance_id"] = data[CONF_ARDUINO_INSTANCE_ID]
+
+    if CONF_ARDUINO_WAIT in data:
+        board_data["arduino_wait"] = data[CONF_ARDUINO_WAIT]
+    if CONF_SLEEP_TUNE in data:
+        board_data["sleep_tune"] = data[CONF_SLEEP_TUNE]
+
+    board_data["autostart"] = False
+    board_data["shutdown_on_exception"] = True
+    board_data["close_loop_on_shutdown"] = False
+
+    board = PymataExpress(**board_data)
+
+    await board.start_aio()
+    return board
+
+
+class FirmataBoardPin(Entity):
+    """Manages a single Firmata board pin."""
+
+    def __init__(self, hass, board_name: str, **kwargs):
+        """Initialize the pin."""
+        self.hass = hass
+        self._name = kwargs[CONF_NAME]
+        self._state = None
+        self._board_name = board_name
+        self._board = hass.data[DOMAIN][self._board_name]
+        self._conf = kwargs
+        self._pin = self._conf[CONF_PIN]
+        self._pin_mode = self._conf[CONF_PIN_MODE]
+        self._firmata_pin_mode = None
+        self._firmata_pin = self._pin
+        if isinstance(self._pin, str):
+            self._pin_type = "analog"
+            self._firmata_pin = int(self._firmata_pin[1:])
+            self._firmata_pin += self._board.api.first_analog_pin
+        else:
+            self._pin_type = "digital"
+        self._location = (DOMAIN, self._board_name, "pin", self._pin)
+        self._unique_id = "_".join(str(i) for i in self._location)
+        self._device_info = self._conf
+        self._device_info.update(
+            {
+                "config_entry_id": self._board.config_entry.entry_id,
+                "pin_type": self._pin_type,
+                "firmata_pin": self._firmata_pin,
+                "firmata_pin_mode": self._firmata_pin_mode,
```

Comment:
This is not valid device info.

Reply:
What constitutes valid device info?

Reply:
https://developers.home-assistant.io/docs/device_registry_index#device-properties